### ISSUE TITLE
support BIP39 mnemonic in descriptors

### ISSUE THREAD

**Issue by jonasschnelli** [bitcoin/bitcoin#19151]
Supporting BIP39 mnemonic in descriptors would allow less painful imports and scans of third party wallet seeds.

I think it could be used almost identical as the existing `xpriv` key.

Example:
`pkh(bip39(room,cross,cube,glance,infant,setup,renew,more,lion,glimpse,retreat,tone,chief,vanish,brisk,destroy,destroy,bounce,knee,analyst,autumn,meadow,divorce,since)/44'/0'/0'/0/0)`
(or slightly different to support the "passphrase")

**Comment by yancyribbens** [comment]
Wouldn't this require at least partial implementation of BIP39 (generating the seed from the mnemonic)?  It seems like there's a long history about why parts of BIP39 have not been included in the wallet to date.

**Comment by RobinLinus** [comment]
Whether we like it or not, BIP39 is the de facto standard for seed backups. Not including it in core only forces people to use less secure wallet software...

**Comment by RobinLinus** [comment]
> Wouldn't this require at least partial implementation of BIP39 (generating the seed from the mnemonic)? It seems like there's a long history about why parts of BIP39 have not been included in the wallet to date.

Whether we like it or not, BIP39 is the de facto standard for seed backups. At this point, not including it in core only forces people to use less secure wallet software...

**Comment by andronoob** [comment]
>It seems like there's a long history about why parts of BIP39 have not been included in the wallet to date.

What I've seen is that, an attempt made by Electrum to solve the alleged problems of BIP39 by [killing BIP39](https://github.com/spesmilo/electrum/issues/1300#issuecomment-112823831) with a newly designed seed format, has failed.

Anyway, **all** seed formats, including but not limited to BIP39, will live **forever**, just like the case of ancient db4.8 wallet.dat files.

(Ditto for WIF privkeys, and some other old incompatible WIF privkeys with different version numbers made by Electrum)

**Comment by andronoob** [comment]
>BIP39 is the de facto standard for seed backups

In fact any wallet app should also take care of ambiguity brought by seed formats like Electrum2.0 and aezeed etc...

I don't think BIP39 is ideal either, I just think the attempts to "fix" this problem usually make the situation even worse in the end.

**Comment by RobinLinus** [comment]
How would you support the passphrase? What do you think about `pkh(bip39([room,cross,cube,glance,infant,setup,renew,more,lion,glimpse,retreat,tone,chief,vanish,brisk,destroy,destroy,bounce,knee,analyst,autumn,meadow,divorce,since],passphrase)/44'/0'/0'/0/0)` ?

**Comment by junderw** [comment]
Support for bip39 is impossible unless we restrict it to English only.

bip39 supports the entirety of the unicode set, normalized with NFKD and decoded into UTF-8 bytes.

afaik, descriptors have a limited set of input characters for the checksum, and therefore can not support BIP39

"Support for a subset of possible BIP39 phrases (English only)" would be more correct.

**Comment by junderw** [comment]
> BIP39 is already english only with its dictionary, I'm pretty sure...

https://github.com/bitcoin/bips/blob/333fc69ab9e6ba6d7ced5236db25f599cd6041b3/bip-0039.mediawiki#wordlist

From the BIP:

> The wordlist can contain native characters, but they must be encoded in UTF-8 using Normalization Form Compatibility Decomposition (NFKD).

Therefore English-only support is technically only supporting a **_subset_** of possible BIP39 phrases.

**Comment by prayank23** [comment]
According to this answer on a website which is considered for lot of issues and PRs in bitcoin core: https://bitcoin.stackexchange.com/a/88244/

BIP 39 is not secure.

Has anything changed? Or using it in descriptors is secure?

Cc: @achow101

**Comment by andronoob** [comment]
@prayank23 
> BIP 39 is not secure.

Nah. BIP39 is secure just like why other seed format is secure - as long as the entropy source is reliable, then the 12/24 words generated using such random source is secure.

The answer you quoted mentioned the weak KDF issue, but it's obvious that KDF has little to no influence on the seed as long as the seed is generated using secure random source. What KDF can have significant influence is only the optional "passphrase", which is usually picked by human (which usually has poor entropy) rather than randomly generated.

-----

The other blamed issue of BIP39 is its lack of version bits (to my understanding it should be called "type" bits more, rather than "version" bits actually).

As there were legacy cases like the Bread (BRD) wallet, BIP39 does deserve some blame: Bread had adopted BIP39 before BIP44 came into being, so that the derivation path BRD picked ironically became "nonstandard" according to BIP44/49/84 etc.

I think Electrum developers do have good intentions. I think they made valid points that users should not worry about something outside the seed itself like the derivation path and corresponding address (output script) type. However I don't think Electrum (2.0) seed has achieved what they advertised - honestly I think Electrum (2.0) seed has on the contrary brought much more chaos the the ecosystem - after Electrum (2.0) seed came into being, the user still can't know WTF the seed phrase at hand is, can't tell it's an Electrum seed, BIP39 seed, or something else - they have been technically ambiguous from the beginning, and even worse, a seed can be probabilistically valid according both the BIP39 and Electrum 2.0 seed validity rules.

**Comment by Sjors** [comment]
I think the context of that Stack Overflow answer is about the wallet Bitcoin Core generates itself. I don't see any objection to _importing_  such wallets.

**Comment by junderw** [comment]
@Sjors the implementation comments still stand though.

Core currently doesn't support 512 bit hashes or PBKDF2 hashing, and adding those just to restore external wallets that we never plan to implement in the core's wallet module... is unnecessary bloat.

**Comment by Sjors** [comment]
The wallet seed is stored as (coerced into) a `CKey` which is constrained to 32 bytes. With descriptor wallets it's probably more doable to add a new seed storage mechanism.

I do think there's a use case for migrating funds from a mnemonic backup to a Bitcoin Core wallet (with or without a sweep). The mnemonics are abundant and there's probably plenty of people who no longer have - or trust - the original wallet that generated it. Right now the only practical way to do that is to import the mnemonic in an arbitrary other (hardware) wallet and then send the coins.

But it's a non trivial change.  Perhaps it's better left to when we have an additional use case for a different seed storage.


That still leaves the possibility of a watch-only import. Arguably this is dangerous and users should instead import an account level xpub if they need a watch-only wallet.

As far hashing, we do have `HMAC-SHA-512`: https://github.com/bitcoin/bitcoin/blob/master/src/crypto/hmac_sha512.h

Assuming this earlier attempt by @mateusz is correct, implementing PBKDF2 pretty short: https://github.com/bitcoin/bitcoin/pull/20361/commits/7e6ba0a3278ff0938c617a234ff07be8a20b06ed#diff-72f3c3f065865085863d0a6577e18effa07a22f3fc2df6e77b5f7089159bc031

**Comment by Sjors** [comment]
Briefly looked at this again. Importing with private key functionality seems doable (not just watch-only).

1. We need to store the extended private key of the root node (easier after #26728): internally our wallet does this by storing the full extended public key and a regular `CKey` for the private key.
2. We don't have to store the mnemonic. But if we wanted to, we have to make sure it gets encrypted along with other private key material.
3. Both seed and passphrase would (initially) be constrained to English, more precisely, these characters (see descriptors.cpp):

```
0123456789()[],'/*abcdefgh@:$%{}
IJKLMNOPQRSTUVWXYZ&+-.;<=>?!^_|~
ijklmnopqrstuvwxyzABCDEFGH`#"\
```

That restriction just needs clear documentation "You can't import […]".

5. The mnemonic (not the password) should be restricted to valid BIP39 (lower case, matching checksum)
6. Add a warning if the length is unusual (not 12 or 24 words)
7. The regular descriptor wallet creation process already uses BIP44/etc derivations. Users can add custom derivations if needed (perhaps by expanding `createwalletdescriptor` after #25907). 
8. PBKDF2 can use existing HMAC-SHA-512 (see comment above)
9. It needs a BIP (in the BIP-380 series)

**Comment by VivaRado** [comment]
Excuse the spam, but here is our implementation of  bip39 with UI in JS, that we use for account access control, client and server side with checksum recalculation. https://github.com/VivaRado/BIP39UI

**Comment by VivaRado** [comment]
So we delete the comment about BIP39 UI implementation because @junderw down voted, without probably even looking at the code. @junderw your behavior is not appreciated. If you do not have something constructive to add, avoid negative displays of grandeur.

**Comment by Sjors** [comment]
#32115 may be useful

**Comment by Sjors** [comment]
Every attempt to do this has been met concept NACKs or lack of review. Maybe we should just decide to not do this and close?

**Comment by ghost** [comment]
Im here because someone swept my issue under the rug and didnt take my concerns seriously, nice to be here. Hope this issue gets resolved. Glad to see there are a few others with common sense, its an important feature beyond knowing how to code.

**Comment by ghost** [comment]
> I really don't see a problem with, or difficulty implementing, support for bip39 in descriptor parsing. Internally it would be mapped to an `xpub` instead, so `listdescriptors` etc wouldn't be able to echo the words back to you, but functionality-wise, that is no different from what [#32115](https://github.com/bitcoin/bitcoin/pull/32115) achieved.

Is that so? So when can we expect to see this feature? Since theres no problems or difficulties?

**Comment by l0rinc** [comment]
@realsetvin, this isn't Twitter, you can't try to bully people into acceptance here.
Please keep it respectful and improve your arguments if you see pushback and try to avoid spamming - otherwise, if you're not listening and not doing the work yourself, you'll likely just get banned anyway.

**Comment by ghost** [comment]
> [@realsetvin](https://github.com/realsetvin), this isn't Twitter, you can't try to bully people into acceptance here. Please keep it respectful and improve your arguments if you see pushback and try to avoid spamming - otherwise, if you're not listening and not doing the work yourself, you'll likely just get banned anyway.

I would give you the same argument, you cant bully people into silence. Imagine the gall to believe you can ban someone from an open source project.

**Comment by caesrcd** [comment]
> Supporting BIP39 mnemonic in descriptors would allow less painful imports and scans of third party wallet seeds.

As I understand it, this would only support importing an existing mnemonic. Would there be an option to randomly generate a BIP-39 mnemonic as well?

**Comment by polespinasa** [comment]
> Would there be an option to randomly generate a BIP-39 mnemonic as well?

No, I don't think so. If ever implemented, will be only importing.
